Fix `execute_block` of cirrus-pallet-executive

liuchengxu opened the pull request:
The main purpose of this PR is to fix the syncing issue between the executors, previously if an executor is connected to another one and tries to sync from it, it'll run into the issue of storage root does not match, which is caused by the `execute_block` missing the step of storage root recording after applying all the extrinsics. More details are expanded in the commit message.

```
2022-05-15 08:48:02 [//Charlie (parachain)] Block prepare storage changes error: Error at calling runtime api: Execution failed: Runtime panicked: Storage root must match that calculated.    
2022-05-15 08:48:02 [//Charlie (parachain)] 💔 Error importing block 0x71d89a3dd61fee4cc5587113e1c0319d2a5f584ba5a3b77263e690e397e9436a: consensus error: Import failed: Error at calling runtime api: Execution failed: Runtime panicked: Storage root must match that calculated.      
```

After this PR, the syncing between the executors works again, and both the traditional syncing and primary block processing are supported now, even though we might don't want to use them at the same time. As long as the executor don't connect to  any other executors, then the syncing can be saved automatically.

liuchengxu committed:
Fix `execute_block` of cirrus-pallet-executive

The sympotom is that previously when we sync a block from the other
executors, an error `storage root does not match` will occur, but the
execution chain is still able to proceed due to we also try to produce a
block based on the primary block in the meanwhile. This patch simply
fixes the syncing issue from other executors.

Basically, we have two approaches to make the execution chain moving
forward:

- Create new execution chain block based on the primary block
- Sync the block from the other executors.

https://github.com/paritytech/substrate/blob/d26123aec44e3df00b921ab0d27cbc1de0b11fa9/client/service/src/client/client.rs#L816
Runtime api `execute_block` is used to process the block synced from the other
executors, while the block produced using the primary block uses the
custom BlockBuilder to execute the block(`apply_extrinsic` +
`finalize_block`). `execute_block` is essentially a combination of
`apply_extrinsic` and `finalize_block`, but the execution trace
recording that was added to `finalize_block` was missing in
`execute_block` before.

liuchengxu committed:
Do not panic if the imported block is already in chain

It's not harmful and allows to connect to the other executors for the
traditional syncing even we might don't want to use it for the effiency.

liuchengxu committed:
Wait for a few more blocks to consume the ready transactions

Probably help this problem https://github.com/subspace/subspace/pull/447#issuecomment-1126890651

nazar-pc reviewed (dismissed):
Looks fine to me, but I don't understand why we need traditional sync on executor at all. I guess it kicks in automatically when connects to bootstrap node?

liuchengxu committed:
Do not connect to other secondary chain node

The test runs well locally, but something is probably still wrong with
the traditional syncing, given that we're not going to support it
anyway, we can just remove it from the test.

liuchengxu committed:
Display the exact error message if an error occurs

So that we can know what was going on in case it fails unexpectedly.

nazar-pc commented on cumulus/client/cirrus-executor/src/tests.rs:
```diff
@@ -40,7 +40,6 @@ async fn test_executor_full_node_catching_up() {
 
 	// run cirrus dave (a secondary chain full node)
 	let dave = cirrus_test_service::TestNodeBuilder::new(tokio_handle, Dave)
-		.connect_to_parachain_node(&charlie)
```
Why was this a problem?

liuchengxu replied:
According to the log https://github.com/subspace/subspace/runs/6441917635?check_suite_focus=true, the node `//Dave (parachain)` somehow resets to the block#0 after a few blocks#5 and is no longer able to move forward, I guess after the syncing issue was fixed, the two ways of producing new execution chain blocks conflicts in some place. But I think it's unnecessary to dig into it since we don't need the traditional syncing anyway. BTW, I have never seen that locally, whether running the tests locally or running a local network.

nazar-pc replied:
We don't need it, but we do have it and will have for a while, so I feel like it just masks an actual problem in tests, but we'll hit it in actual deployment sooner or later anyway.

Reproduction locally might be easier when machine is under load, try to run tests while running something heavy in parallel, like OpenSSL benchmark (`openssl speed -multi 24`).

liuchengxu replied:
Haven't found anything new running tests with a busy machine locally.

You won't have this problem as long as you don't connect to the other executors, only syncing from other executors is affected, which is broken already even worse before this PR.

Unless you want to support the traditional syncing between the executors officially, I don't think we need to invest more time debugging the issue, we already have enough for the final goal.

liuchengxu replied:
The initial motivation for this PR is that we spent quite some time debugging the syncing issue between the executors as we don't know what's the root cause and if it could affect the other things. Since we figured it out and the fix is simple, I made this PR, without foreseeing the test issue as I have never seen it before. But if it requires more time to fix it properly, it's not profitable if it will not be supported officially IMHO. Hope this explanation can solve some of your confusion :D

nazar-pc replied:
As I said, things like this might shoot later and it'd be really nice to know what the root cause is regardless.

liuchengxu replied:
Less likely if this piece of code will not be enabled in practice. I can make it a backlog item and try debugging when our schedule is not pretty tight.

nazar-pc replied:
I think that is a good idea
